Commit ID: 38d5fc4e5c686d8ba0af0d9bb8aaf3cf10b215f6
Change ID: wyvkquvwmttmqxtulsntkyxltlpznmxq
Bookmarks: fritz-ipv6-prefix fritz-ipv6-prefix@git fritz-ipv6-prefix@origin
Author   : Marie Ramlow <me@nycode.dev> (2026-04-02 17:40:08)
Committer: Marie Ramlow <me@nycode.dev> (2026-05-23 14:23:55)

    Add IPv6 prefix sensor to FRITZ!Box Tools

diff --git a/homeassistant/components/fritz/icons.json b/homeassistant/components/fritz/icons.json
index 3d220a96e7..efd7ee7628 100644
--- a/homeassistant/components/fritz/icons.json
+++ b/homeassistant/components/fritz/icons.json
@@ -29,6 +29,9 @@
       "gb_sent": {
         "default": "mdi:upload"
       },
+      "ipv6_prefix": {
+        "default": "mdi:earth"
+      },
       "kb_s_received": {
         "default": "mdi:download"
       },
diff --git a/homeassistant/components/fritz/sensor.py b/homeassistant/components/fritz/sensor.py
index b2290ade9a..1c0ad83a3f 100644
--- a/homeassistant/components/fritz/sensor.py
+++ b/homeassistant/components/fritz/sensor.py
@@ -62,6 +62,12 @@
     return str(status.external_ipv6)
 
 
+def _retrieve_ipv6_prefix_state(status: FritzStatus, last_value: str) -> str:
+    """Return ipv6 prefix from device."""
+    info = status.ipv6_prefix_info
+    return f"{info['NewIPv6Prefix']}/{info['NewPrefixLength']}"
+
+
 def _retrieve_kb_s_sent_state(status: FritzStatus, last_value: str) -> float:
     """Return upload transmission rate."""
     return round(status.transmission_rate[0] / 1000, 1)  # type: ignore[no-any-return]
@@ -184,6 +190,12 @@
         is_suitable=lambda info: info.ipv6_active,
     ),
     FritzConnectionSensorEntityDescription(
+        key="ipv6_prefix",
+        translation_key="ipv6_prefix",
+        value_fn=_retrieve_ipv6_prefix_state,
+        is_suitable=lambda info: info.ipv6_active,
+    ),
+    FritzConnectionSensorEntityDescription(
         key="connection_uptime",
         translation_key="connection_uptime",
         device_class=SensorDeviceClass.UPTIME,
diff --git a/homeassistant/components/fritz/strings.json b/homeassistant/components/fritz/strings.json
index fb269118aa..94f01f534d 100644
--- a/homeassistant/components/fritz/strings.json
+++ b/homeassistant/components/fritz/strings.json
@@ -132,6 +132,9 @@
       "gb_sent": {
         "name": "GB sent"
       },
+      "ipv6_prefix": {
+        "name": "IPv6 prefix"
+      },
       "kb_s_received": {
         "name": "Download throughput"
       },
diff --git a/tests/components/fritz/const.py b/tests/components/fritz/const.py
index 4415f563ff..15deba8d07 100644
--- a/tests/components/fritz/const.py
+++ b/tests/components/fritz/const.py
@@ -149,6 +149,10 @@
         },
         "GetExternalIPAddress": {"NewExternalIPAddress": "1.2.3.4"},
         "X_AVM_DE_GetExternalIPv6Address": {"NewExternalIPv6Address": "fec0::1"},
+        "X_AVM_DE_GetIPv6Prefix": {
+            "NewIPv6Prefix": "2001:db8::",
+            "NewPrefixLength": 32,
+        },
     },
     "WANPPPConnection1": {
         "GetInfo": {
diff --git a/tests/components/fritz/snapshots/test_sensor.ambr b/tests/components/fritz/snapshots/test_sensor.ambr
index 4df192c540..16f3abf794 100644
--- a/tests/components/fritz/snapshots/test_sensor.ambr
+++ b/tests/components/fritz/snapshots/test_sensor.ambr
@@ -324,6 +324,56 @@
     'state': '1.7',
   })
 # ---
+# name: test_sensor_cpu_temp_not_supported[None-return_values1][sensor.mock_title_ipv6_prefix-entry]
+  EntityRegistryEntrySnapshot({
+    'aliases': list([
+      None,
+    ]),
+    'area_id': None,
+    'capabilities': None,
+    'config_entry_id': <ANY>,
+    'config_subentry_id': <ANY>,
+    'device_class': None,
+    'device_id': <ANY>,
+    'disabled_by': None,
+    'domain': 'sensor',
+    'entity_category': None,
+    'entity_id': 'sensor.mock_title_ipv6_prefix',
+    'has_entity_name': True,
+    'hidden_by': None,
+    'icon': None,
+    'id': <ANY>,
+    'labels': set({
+    }),
+    'name': None,
+    'object_id_base': 'IPv6 prefix',
+    'options': dict({
+    }),
+    'original_device_class': None,
+    'original_icon': None,
+    'original_name': 'IPv6 prefix',
+    'platform': 'fritz',
+    'previous_unique_id': None,
+    'suggested_object_id': None,
+    'supported_features': 0,
+    'translation_key': 'ipv6_prefix',
+    'unique_id': '1CED6F123411-ipv6_prefix',
+    'unit_of_measurement': None,
+  })
+# ---
+# name: test_sensor_cpu_temp_not_supported[None-return_values1][sensor.mock_title_ipv6_prefix-state]
+  StateSnapshot({
+    'attributes': ReadOnlyDict({
+      'friendly_name': 'Mock Title IPv6 prefix',
+    }),
+    'context': <ANY>,
+    'entity_id': 'sensor.mock_title_ipv6_prefix',
+    'last_changed': <ANY>,
+    'last_reported': <ANY>,
+    'last_updated': <ANY>,
+    'state': '2001:db8::/32',
+  })
+# ---
 # name: test_sensor_cpu_temp_not_supported[None-return_values1][sensor.mock_title_link_download_noise_margin-entry]
   EntityRegistryEntrySnapshot({
     'aliases': list([
@@ -1182,6 +1232,56 @@
     'state': '1.7',
   })
 # ---
+# name: test_sensor_cpu_temp_not_supported[None-return_values2][sensor.mock_title_ipv6_prefix-entry]
+  EntityRegistryEntrySnapshot({
+    'aliases': list([
+      None,
+    ]),
+    'area_id': None,
+    'capabilities': None,
+    'config_entry_id': <ANY>,
+    'config_subentry_id': <ANY>,
+    'device_class': None,
+    'device_id': <ANY>,
+    'disabled_by': None,
+    'domain': 'sensor',
+    'entity_category': None,
+    'entity_id': 'sensor.mock_title_ipv6_prefix',
+    'has_entity_name': True,
+    'hidden_by': None,
+    'icon': None,
+    'id': <ANY>,
+    'labels': set({
+    }),
+    'name': None,
+    'object_id_base': 'IPv6 prefix',
+    'options': dict({
+    }),
+    'original_device_class': None,
+    'original_icon': None,
+    'original_name': 'IPv6 prefix',
+    'platform': 'fritz',
+    'previous_unique_id': None,
+    'suggested_object_id': None,
+    'supported_features': 0,
+    'translation_key': 'ipv6_prefix',
+    'unique_id': '1CED6F123411-ipv6_prefix',
+    'unit_of_measurement': None,
+  })
+# ---
+# name: test_sensor_cpu_temp_not_supported[None-return_values2][sensor.mock_title_ipv6_prefix-state]
+  StateSnapshot({
+    'attributes': ReadOnlyDict({
+      'friendly_name': 'Mock Title IPv6 prefix',
+    }),
+    'context': <ANY>,
+    'entity_id': 'sensor.mock_title_ipv6_prefix',
+    'last_changed': <ANY>,
+    'last_reported': <ANY>,
+    'last_updated': <ANY>,
+    'state': '2001:db8::/32',
+  })
+# ---
 # name: test_sensor_cpu_temp_not_supported[None-return_values2][sensor.mock_title_link_download_noise_margin-entry]
   EntityRegistryEntrySnapshot({
     'aliases': list([
@@ -2040,6 +2140,56 @@
     'state': '1.7',
   })
 # ---
+# name: test_sensor_cpu_temp_not_supported[side_effect0-None][sensor.mock_title_ipv6_prefix-entry]
+  EntityRegistryEntrySnapshot({
+    'aliases': list([
+      None,
+    ]),
+    'area_id': None,
+    'capabilities': None,
+    'config_entry_id': <ANY>,
+    'config_subentry_id': <ANY>,
+    'device_class': None,
+    'device_id': <ANY>,
+    'disabled_by': None,
+    'domain': 'sensor',
+    'entity_category': None,
+    'entity_id': 'sensor.mock_title_ipv6_prefix',
+    'has_entity_name': True,
+    'hidden_by': None,
+    'icon': None,
+    'id': <ANY>,
+    'labels': set({
+    }),
+    'name': None,
+    'object_id_base': 'IPv6 prefix',
+    'options': dict({
+    }),
+    'original_device_class': None,
+    'original_icon': None,
+    'original_name': 'IPv6 prefix',
+    'platform': 'fritz',
+    'previous_unique_id': None,
+    'suggested_object_id': None,
+    'supported_features': 0,
+    'translation_key': 'ipv6_prefix',
+    'unique_id': '1CED6F123411-ipv6_prefix',
+    'unit_of_measurement': None,
+  })
+# ---
+# name: test_sensor_cpu_temp_not_supported[side_effect0-None][sensor.mock_title_ipv6_prefix-state]
+  StateSnapshot({
+    'attributes': ReadOnlyDict({
+      'friendly_name': 'Mock Title IPv6 prefix',
+    }),
+    'context': <ANY>,
+    'entity_id': 'sensor.mock_title_ipv6_prefix',
+    'last_changed': <ANY>,
+    'last_reported': <ANY>,
+    'last_updated': <ANY>,
+    'state': '2001:db8::/32',
+  })
+# ---
 # name: test_sensor_cpu_temp_not_supported[side_effect0-None][sensor.mock_title_link_download_noise_margin-entry]
   EntityRegistryEntrySnapshot({
     'aliases': list([
@@ -2898,6 +3048,56 @@
     'state': '1.7',
   })
 # ---
+# name: test_sensor_cpu_temp_not_supported[side_effect3-None][sensor.mock_title_ipv6_prefix-entry]
+  EntityRegistryEntrySnapshot({
+    'aliases': list([
+      None,
+    ]),
+    'area_id': None,
+    'capabilities': None,
+    'config_entry_id': <ANY>,
+    'config_subentry_id': <ANY>,
+    'device_class': None,
+    'device_id': <ANY>,
+    'disabled_by': None,
+    'domain': 'sensor',
+    'entity_category': None,
+    'entity_id': 'sensor.mock_title_ipv6_prefix',
+    'has_entity_name': True,
+    'hidden_by': None,
+    'icon': None,
+    'id': <ANY>,
+    'labels': set({
+    }),
+    'name': None,
+    'object_id_base': 'IPv6 prefix',
+    'options': dict({
+    }),
+    'original_device_class': None,
+    'original_icon': None,
+    'original_name': 'IPv6 prefix',
+    'platform': 'fritz',
+    'previous_unique_id': None,
+    'suggested_object_id': None,
+    'supported_features': 0,
+    'translation_key': 'ipv6_prefix',
+    'unique_id': '1CED6F123411-ipv6_prefix',
+    'unit_of_measurement': None,
+  })
+# ---
+# name: test_sensor_cpu_temp_not_supported[side_effect3-None][sensor.mock_title_ipv6_prefix-state]
+  StateSnapshot({
+    'attributes': ReadOnlyDict({
+      'friendly_name': 'Mock Title IPv6 prefix',
+    }),
+    'context': <ANY>,
+    'entity_id': 'sensor.mock_title_ipv6_prefix',
+    'last_changed': <ANY>,
+    'last_reported': <ANY>,
+    'last_updated': <ANY>,
+    'state': '2001:db8::/32',
+  })
+# ---
 # name: test_sensor_cpu_temp_not_supported[side_effect3-None][sensor.mock_title_link_download_noise_margin-entry]
   EntityRegistryEntrySnapshot({
     'aliases': list([
@@ -3814,6 +4014,56 @@
     'state': '1.7',
   })
 # ---
+# name: test_sensor_setup[sensor.mock_title_ipv6_prefix-entry]
+  EntityRegistryEntrySnapshot({
+    'aliases': list([
+      None,
+    ]),
+    'area_id': None,
+    'capabilities': None,
+    'config_entry_id': <ANY>,
+    'config_subentry_id': <ANY>,
+    'device_class': None,
+    'device_id': <ANY>,
+    'disabled_by': None,
+    'domain': 'sensor',
+    'entity_category': None,
+    'entity_id': 'sensor.mock_title_ipv6_prefix',
+    'has_entity_name': True,
+    'hidden_by': None,
+    'icon': None,
+    'id': <ANY>,
+    'labels': set({
+    }),
+    'name': None,
+    'object_id_base': 'IPv6 prefix',
+    'options': dict({
+    }),
+    'original_device_class': None,
+    'original_icon': None,
+    'original_name': 'IPv6 prefix',
+    'platform': 'fritz',
+    'previous_unique_id': None,
+    'suggested_object_id': None,
+    'supported_features': 0,
+    'translation_key': 'ipv6_prefix',
+    'unique_id': '1CED6F123411-ipv6_prefix',
+    'unit_of_measurement': None,
+  })
+# ---
+# name: test_sensor_setup[sensor.mock_title_ipv6_prefix-state]
+  StateSnapshot({
+    'attributes': ReadOnlyDict({
+      'friendly_name': 'Mock Title IPv6 prefix',
+    }),
+    'context': <ANY>,
+    'entity_id': 'sensor.mock_title_ipv6_prefix',
+    'last_changed': <ANY>,
+    'last_reported': <ANY>,
+    'last_updated': <ANY>,
+    'state': '2001:db8::/32',
+  })
+# ---
 # name: test_sensor_setup[sensor.mock_title_link_download_noise_margin-entry]
   EntityRegistryEntrySnapshot({
     'aliases': list([
